assert/assertions.go

```diff
@@ -1591,3 +1591,36 @@ func Eventually(t TestingT, condition func() bool, waitFor time.Duration, tick t
 		}
 	}
 }
+
+// Never asserts that the given condition doesn't satisy in waitFor time,
+// periodically checking target function each tick.
+//
+//    assert.Never(t, func() bool { return false; }, time.Second, 10*time.Millisecond)
+func Never(t TestingT, condition func() bool, waitFor time.Duration, tick time.Duration, msgAndArgs ...interface{}) bool {
+	if h, ok := t.(tHelper); ok {
+		h.Helper()
+	}
+
+	ch := make(chan bool, 1)
+
+	timer := time.NewTimer(waitFor)
+	defer timer.Stop()
+
+	ticker := time.NewTicker(tick)
+	defer ticker.Stop()
+
+	for tick := ticker.C; ; {
+		select {
+		case <-timer.C:
+			return true
+		case <-tick:
+			tick = nil
+			go func() { ch <- condition() }()
```

Comment:
If I am not mistaken, will this not end up spawning _timer_/_tick_ goroutines of the provided function? Should we not rather run the `condition()` function once and only do checks here?

Reply:
What do you mean by running the condition function only once? I.e. where would you place that in the code, it is not clear to me. 

To be honest I just copied code from the existing `Eventually` function.  Here's the place where that part of the code was touched last: https://github.com/stretchr/testify/commit/f1bd0923b8320ddfb0a8a9d919b5aeb8b7a4cc40. The previous place was when the code was added. It looks like it has been done like this since the beginning.

Reply:
To clarify what I meant: It looks to me that on every timer tick, a goroutine is launched triggers calls the condition function. I'm not 100% sure if that's the case, but it does look like it to me. This means that we're re-running the function on every tick which would prove to be very inefficient and with possibly adverse side-effects if it's not a pure function/stateless.

> To be honest I just copied code from the existing Eventually function. Here's the place where that part of the code was touched last: f1bd092. The previous place was when the code was added. It looks like it has been done like this since the beginning.

Makes sense, maybe it's worth adding a unit test to check exactly that the existing `Eventually` logic doesn't call multiple copies of the same condition function. Thoughts?

I'd be happy to help where necessary. If you wish I could add the test for the existing, `Evenntually` function and we could work from there given the outcome?

Reply:
From my understanding, rerunning the function every tick is the desired thing to do here. The function that's passed in is expected to not be pure, otherwise neither `Never` or `Eventually` would make any sense. If the functions are pure, `assert.Never` would be equivalent to `assert.False` and `assert.Eventually` would be equivalent to `assert.True`.

Reply:
:thinking: 

I think my understanding may be a bit off. Is the point of `Eventually` and `Never` to assert that a function completes/doesn't complete within a certain amount of time?

Reply:
No, that seems to be a misunderstanding on your part.

`Eventually` asserts that the given function returns true at some point before the timeout
`Never` asserts that the given function _does not_ return true at some point before the timeout

Reply:
Got you. I'm still unsure of this but I'm happy to approve and commit. I'll spend some time figuring this out and maybe put out a refactoring PR; which I'd appreciate your feedback on :)

Happy for you to merge this and thank you for your contribution :)

Reply:
Sure, just ping me if there's something you want me to look at! Thanks for taking a look at this.